### DISCUSSION TITLE
Instructions to use Codeql

### DISCUSSION THREAD

**Discussion by AhmadALBarqawi** [github/codeql#14259] [Q&A]
Hello guys, I am new to codeql and would like to use it in order to get analysis reports of individual python files, I am trying to see what and how many CWEs are in each python file. Can you guys provide me with information on how I would be able to create databases for these files (which are located locally on my computer)? and also if there are already queries setup to use against these codebases to check for all python CWEs that are detectable by codeql?

**Comment by aibaars** [comment]
The documentation can be found at https://codeql.github.com/docs/ . For local use the [CodeQL CLI](https://codeql.github.com/docs/codeql-cli) section is a good starting point. The main commands you need are `codeql database create` and `codeql database analyze`. For your use-case I'd recommend the `python-security-extended` query set.

**Reply by AhmadALBarqawi** [reply]
Sounds great Arthur, appreciate your help. I went deeper into understanding these commands and I believe I got something here. However, I am looking into actually getting the human interpretable results now and I am facing some troubles. Do I need to decode every single .bqrs file? and why are the outputs not automatically in the output csv file I specified in the analyze command?

**Reply by aibaars** [reply]
The outputs should normally be in the generated CSV file. Do you see any interesting results in the bqrs files that are not in the CSV? Note that the CSV file won't contain any results outside the "source root". It could happen that CodeQL finds an alert in for example the python standard library while it is analysing your code, you probably only care about your own code which is why `codeql database analyze` only reports the alert in your source folder. I think `codeql database analyze` prints info messages about dropped results if you run it with a higher verbosity level.

**Reply by AhmadALBarqawi** [reply]
The problem is that the csv file is completely empty after I run the analyze command. Does the analyze command create its' own csv file or do I have to specify an already created one for the output?

**Reply by aibaars** [reply]
The command creates its own CSV file. Could you share one of your file that you think contains a vulnerability that CodeQL should find?

**Reply by AhmadALBarqawi** [reply]
Sorry for the late reply, I got back to looking into my problem and your intuition was correct, I changed the the file I was analyzing and now I am getting results. I am now looking into the CWEs that are covered by CodeQL for python in the queries folder and they are much less than the ones on CWEs coverage webpage. How can I get the rest?

**Reply by aibaars** [reply]
What do you mean by the "queries" folder? The CWE table on https://codeql.github.com/codeql-query-help/python-cwe/ contains the query identifiers. You should be able to find all of them in https://github.com/github/codeql/tree/main/python/ql/src . Some of them might be "experimental" and not part of the set of queries that get run by default.

**Reply by AhmadALBarqawi** [reply]
Exactly, I can't find all the CWEs mentioned here https://codeql.github.com/codeql-query-help/python-cwe/ in the folder https://github.com/github/codeql/tree/main/python/ql/src. An example would be CWE-668.

**Reply by AhmadALBarqawi** [reply]
Another question I have is whether I can change the output of the csv file to also include the CWE name and number?

**Reply by aibaars** [reply]
I don't think the CSV output contains that information, however, the SARIF output (`--format=sarifv2.1.0`) includes all that information.

**Reply by AhmadALBarqawi** [reply]
Thank you for that, using the sarif format is definitely much better. My earlier question remains as I am unable to find a folder in the source files for each CWE supposedly covered by CodeQL.
